examples/layers/services/isolate.dart

```diff
@@ -0,0 +1,308 @@
+  : _receivePort = new ReceivePort() {
+    _receivePort.listen(_handleMessage);
+  }
+
+  CalculationState _state = CalculationState.idle;
+  CalculationState get state => _state;
+  bool get isRunning => _state != CalculationState.idle;
+
+  double _completed = 0.0;
+  double _total = 1.0;
+
+  final OnProgressListener onProgressListener;
+  final OnResultListener onResultListener;
+
+  /// Start the background computation.
+  ///
+  /// Does nothing if the computation is already running.
+  void start() {
+    if (!isRunning) {
+      _state = CalculationState.loading;
+      _runCalculation();
+    }
+  }
+
+  /// Stop the background computation.
+  ///
+  /// Kills the [Isolate] immediately, if spawned. Does nothing if the
+  /// computation is not running.
+  void stop() {
+    if (isRunning) {
+      _state = CalculationState.idle;
+      if (_isolate != null) {
+        _isolate.kill(priority: Isolate.IMMEDIATE);
+        _isolate = null;
+        _completed = 0.0;
+        _total = 1.0;
+      }
+    }
+  }
+
+  final ReceivePort _receivePort;
+  Isolate _isolate;
+
+  void _runCalculation() {
+    // Load the JSON string. Note that this is done in the main isolate; at the
+    // moment, spawned isolates do not have access to Mojo services, including
+    // the root bundle (see https://github.com/flutter/flutter/issues/3294).
+    // However, the loading process is asynchronous, so the UI will not block
+    // while the file is loaded.
+    rootBundle.loadString('services/data.json').then((String data) {
+      if (isRunning) {
+        List<dynamic> message = [ _receivePort.sendPort, data ];
+        // Spawn an isolate to JSON-parse the file contents. The JSON parsing
+        // is synchronous, so if done in the main isolate, the UI would block.
+        Isolate.spawn(_calculate, message).then((Isolate isolate) {
+          _state = CalculationState.calculating;
+          if (!isRunning) {
+            isolate.kill(priority: Isolate.IMMEDIATE);
+          } else {
+            _isolate = isolate;
+          }
+        });
+      }
+    });
+  }
+
+  void _handleMessage(dynamic message) {
+    if (message is List<double>) {
+      _completed = message[0];
+      _total = message[1];
+      if (onProgressListener != null)
+        onProgressListener(_completed, _total);
+    } else if (message is String) {
+      _completed = 0.0;
+      _total = 1.0;
+      _isolate = null;
+      _state = CalculationState.idle;
+      if (onResultListener != null)
+        onResultListener(message);
+    }
+  }
+
+  /// Main entry point for the spawned [Isolate].
+  ///
+  /// This entry point must be static, and its (single) argument must match
+  /// the message passed in `Isolate.spawn` above. Typically, some part of the
+  /// message will contain a [SendPort] so that the spawned isolate can
+  /// communicate back to the main isolate.
+  static void _calculate(List<dynamic> message) {
```

Comment:
Add a comment about what the values of global variables will be in this context. Are they all null? default values? If there's a static initialised by a function, does the function run again?